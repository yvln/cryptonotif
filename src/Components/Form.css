
.Form {
    background-color: #474044;
    max-width: 700px;
    margin: 0 auto;
  }
  