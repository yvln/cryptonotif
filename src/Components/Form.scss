.Form {
  text-align: center;

  input,
  select {
    background-color: #e8edef;
    border: none;
    padding: 5px;
  }

  // override default style of Autosuggest component
  .react-autosuggest__input {
    z-index: 2;
  }

  .react-autosuggest__container--open {
    color: black;

    > .react-autosuggest__suggestions-container {
      margin: 0 auto;

      left: 50%;
      transform: translateX(-50%);

      background-color: rgba(255, 255, 255, 0.9);

      > ul {
        margin: 0;
        padding: 0;

        > li {
          list-style-type: none;
          font-size: 10px;
          text-align: left;

          &:hover {
            background-color: rgba(0, 0, 0, 0.2);
          }

          &:not(:last-child) {
            border-bottom: 1px solid rgba(255, 105, 120, 0.7);
          }
        }
      }
    }
  }
}

.FormCreate {
  padding: 20px;
  max-width: 300px;
  background-color: #e8edef;
  border: 5px solid #264653;
  font-size: 20px;
  margin: 50px auto;

  > * {
    margin-bottom: 10px;
  }

  .Button {
    height: 50px;
    width: 50px;
    border-radius: 50%;
    vertical-align: center;
    font-size: 18px;
    margin-top: 20px;
    background-color: #264653;
    color: white;

    &:hover {
      border-bottom: 0;
    }
  }

  input {
    height: 20px;
    border-bottom: 2px solid rgb(255, 105, 120);
  }

  select {
    height: 30px;
    background-color: #6d435a;
    color: white;
  }

  // override default style of Autosuggest component
  .react-autosuggest__input {
    width: 150px;
  }

  .react-autosuggest__container--open {
    position: relative;

    > .react-autosuggest__suggestions-container {
      position: absolute;
      line-height: 2em;
      width: 160px;

      > ul > li {
        padding-left: 30px;
      }
    }
  }

}

.FormEdit {
  > * {
    margin-bottom: 5px;
  }

  // override default style of Autosuggest component
  .react-autosuggest__input {
    width: 70px;
  }

  .react-autosuggest__container--open {
    position: relative;

    > .react-autosuggest__suggestions-container {
      position: absolute;
      line-height: 1.5em;
      width: 80px;

      > ul > li {
        padding-left: 10px;
      }
    }
  }
}
