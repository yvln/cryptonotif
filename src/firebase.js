/* eslint-disable */

const firebase = require('firebase');
require('dotenv').config();

const config = {
  apiKey: process.env.API_KEY,
  authDomain: 'cryptonotif-56a23.firebaseapp.com',
  databaseURL: 'https://cryptonotif-56a23.firebaseio.com',
  projectId: 'cryptonotif-56a23',
  storageBucket: 'cryptonotif-56a23.appspot.com',
  messagingSenderId: '1008924569797',
};

export const initializeFirebase = () => {
  firebase.initializeApp(config);
  return true;
};

export const getDatabase = () => {
  return firebase.database();
};
