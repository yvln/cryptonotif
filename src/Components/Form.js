import React, { Component } from "react";
import Autosuggest from "react-autosuggest";

import { getDatabase } from "../firebase";
import "./Form.scss";

let suggestions;
const url = "https://rest.coinapi.io/v1/assets";
const apikey = "6F3ABC62-B928-4F74-8194-24EA076F89C8";

class Form extends Component {
  state = {
    email: "",
    error: undefined,
    suggestions: [],
    action: "under",
    amount: "",
    asset: "",
    currency: "USD"
  };

  componentDidMount() {
    if (this.props.initialValues) {
      const {
        action,
        amount,
        asset,
        currency,
        email
      } = this.props.initialValues.data;

      this.setState({
        action,
        amount,
        asset,
        currency,
        email
      });
    }

    fetch(url, {
      headers: {
        "X-CoinAPI-Key": apikey
      }
    })
      .then(res =>
        res.json().then(response => {
          suggestions = response
            .filter(asset => asset.type_is_crypto === 1)
            .map(asset => {
              return {
                label: asset.asset_id
              };
            });
        })
      )
      .catch(err => console.log(err));
  }

  addOrEditAlert = e => {
    e.preventDefault();
    const db = getDatabase();
    const { action, amount, asset, currency, email } = this.state;
    const key = `${action}${amount}${currency}${asset}${email
      .split(".")
      .join("")}`;
    const data = {
      action,
      asset,
      amount,
      currency,
      email,
      emailSent: false,
      key,
    };

    db.ref(`/alert/${key}`)
      .once("value")
      .then(snapshot => {
        if (snapshot.val()) {
          return this.setState({
            error: "This alert has already been created."
          });
        }
      });

    this.setState({
      error: undefined
    });

    if (this.props.initialValues) {
      return db.ref().update({ [`/alert/${key}`]: data });
    }

    return db.ref(`/alert/${key}`).set(data);
  };

  getSuggestions = value => {
    const inputValue = value.trim().toUpperCase();
    return !inputValue.length
      ? []
      : suggestions.filter(suggestion => suggestion.label.includes(inputValue));
  };

  getSuggestionValue = value => value.label;

  onChangeInput = (e, key) => {
    this.setState({
      [key]: e.target.value
    });
  };

  onChangeAsset = (e, { newValue }) => {
    this.setState({
      asset: newValue
    });
  };

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: []
    });
  };

  onSuggestionsFetchRequested = ({ value }) => {
    this.setState({
      suggestions: this.getSuggestions(value)
    });
  };

  renderSuggestion = suggestion => <div>{suggestion.label}</div>;

  render() {
    const { email, action, amount, asset, currency, suggestions } = this.state;
    const inputProps = {
      placeholder: "BTC, ETH, ...",
      value: asset,
      onChange: this.onChangeAsset
    };

    return (
      <form
        className={`Form ${
          this.props.initialValues ? "FormEdit" : "FormCreate"
        }`}
        onSubmit={this.addOrEditAlert}
      >
        {!this.props.initialValues && this.state.error && (
          <div className="error">{this.state.error}</div>
        )}

        <div>Alert me when</div>
        <Autosuggest
          suggestions={suggestions.splice(0, 9)}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          getSuggestionValue={this.getSuggestionValue}
          renderSuggestion={this.renderSuggestion}
          inputProps={inputProps}
        />
        <select
          value={action}
          onChange={e => this.onChangeInput(e, "action")}
          name="action"
          className="menuAction"
        >
          <option value="under">falls under</option>
          <option value="above">is above</option>
        </select>
        <div>
          <input
            onChange={e => this.onChangeInput(e, "amount")}
            type="number"
            name="amount"
            value={amount}
          />

          <select
            value={currency}
            onChange={e => this.onChangeInput(e, "currency")}
            name="currency"
            className="menuCurrency"
          >
            <option value="USD">USD</option>
            <option value="EUR">EUR</option>
          </select>
        </div>
        <div>
          <span>to </span>
          <input
            onChange={e => this.onChangeInput(e, "email")}
            placeholder="your email"
            name="email"
            disabled={this.props.initialValues}
            type="email"
            value={email}
          />
          {this.props.initialValues &&
            "To update the email address, please create another alert."}
        </div>
        <input className="Button" type="submit" value="GO" />
      </form>
    );
  }
}

export default Form;
