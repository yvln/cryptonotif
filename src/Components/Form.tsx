import React, { Component } from 'react';
import Autosuggest from 'react-autosuggest';

import { getDatabase } from '../firebase';
import './Form.scss';

const url = 'https://rest.coinapi.io/v1/assets';
const apikey = '6F3ABC62-B928-4F74-8194-24EA076F89C8';

type FormProps = {
  closeForm?: () => void;
  initialValues?: {
    data: {
      action: string;
      amount: string;
      asset: string;
      currency: string;
      email: string;
      key: string;
    };
  };
};

type InputKey = 'action' | 'amount' | 'currency' | 'email';

type Suggestion = {
  label: string;
};

type ResponseCoinAPI = {
  asset_id: string;
  data_end: Date;
  data_orderbook_end: Date;
  data_orderbook_start: Date;
  data_quote_end: Date;
  data_quote_start: Date;
  data_start: Date;
  data_symbols_count: number;
  data_trade_count: number;
  data_trade_end: Date;
  data_trade_start: Date;
  name: string;
  type_is_crypto: number;
};

type FormState = {
  error: string | undefined;
  suggestions: Suggestion[];
  asset: string;
} & {
  // FIX: update type to be more specific
  // [key in InputKey]: string;
  [key: string]: any;
};

let suggestions: Suggestion[];

class Form extends Component<FormProps, FormState> {
  state = {
    email: '',
    error: undefined,
    suggestions: [],
    action: 'under',
    amount: '',
    asset: '',
    currency: 'USD',
  };

  componentDidMount() {
    if (this.props.initialValues) {
      const {
        action,
        amount,
        asset,
        currency,
        email,
      } = this.props.initialValues.data;

      this.setState({
        action,
        amount,
        asset,
        currency,
        email,
      });
    }

    fetch(url, {
      headers: {
        'X-CoinAPI-Key': apikey,
      },
    }).then(res =>
      res.json().then((response: ResponseCoinAPI[]) => {
        suggestions = response
          .filter(asset => asset.type_is_crypto === 1)
          .map(asset => {
            return {
              label: asset.asset_id,
            };
          });
      })
    );
  }

  addOrEditAlert = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const db = getDatabase();
    const { action, amount, asset, currency, email } = this.state;
    const key = `${action}${amount}${currency}${asset}${email
      .split('.')
      .join('')}`;
    const data = {
      action,
      asset,
      amount,
      currency,
      email,
      emailSent: false,
      key,
    };

    db
      .ref(`/alert/${key}`)
      .once('value')
      .then(snapshot => {
        if (snapshot.val()) {
          return this.setState({
            error: 'This alert has already been created.',
          });
        }
      });

    this.setState({
      error: undefined,
    });

    if (this.props.initialValues) {
      return db.ref().update({ [`/alert/${key}`]: data });
    }

    return db.ref(`/alert/${key}`).set(data);
  };

  getSuggestions = (value: string) => {
    const inputValue = value.trim().toUpperCase();
    return !inputValue.length
      ? []
      : suggestions.filter((suggestion: Suggestion) =>
          suggestion.label.includes(inputValue)
        );
  };

  getSuggestionValue = (value: Suggestion) => value.label;

  onChangeInput = (
    e: React.FormEvent<HTMLInputElement> | React.ChangeEvent<HTMLSelectElement>,
    key: InputKey
  ) => {
    this.setState({
      [key]: e.currentTarget.value,
    });
  };

  onChangeAsset = (
    e: React.FormEvent<HTMLInputElement>,
    { newValue }: { newValue: string }
  ) => {
    this.setState({
      asset: newValue,
    });
  };

  onSuggestionsClearRequested = () => {
    this.setState({
      suggestions: [],
    });
  };

  onSuggestionsFetchRequested = ({ value }: { value: string }) => {
    this.setState({
      suggestions: this.getSuggestions(value),
    });
  };

  renderSuggestion = (suggestion: Suggestion) => <div>{suggestion.label}</div>;

  render() {
    const { email, action, amount, asset, currency, suggestions } = this.state;
    const inputProps = {
      placeholder: 'BTC, ETH, ...',
      value: asset,
      onChange: this.onChangeAsset,
    };

    return (
      <form
        className={`Form ${
          this.props.initialValues ? 'FormEdit' : 'FormCreate'
        }`}
        onSubmit={this.addOrEditAlert}
      >
        {!this.props.initialValues &&
          this.state.error && <div className="error">{this.state.error}</div>}

        <div>Alert me when</div>
        <Autosuggest
          suggestions={suggestions.splice(0, 9)}
          onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={this.onSuggestionsClearRequested}
          getSuggestionValue={this.getSuggestionValue}
          renderSuggestion={this.renderSuggestion}
          inputProps={inputProps}
        />
        <select
          value={action}
          onChange={e => this.onChangeInput(e, 'action')}
          name="action"
          className="menuAction"
        >
          <option value="under">falls under</option>
          <option value="above">is above</option>
        </select>
        <div>
          <input
            onChange={e => this.onChangeInput(e, 'amount')}
            type="number"
            name="amount"
            value={amount}
          />

          <select
            value={currency}
            onChange={e => this.onChangeInput(e, 'currency')}
            name="currency"
            className="menuCurrency"
          >
            <option value="USD">USD</option>
            <option value="EUR">EUR</option>
          </select>
        </div>
        <div>
          <span>to </span>
          <input
            onChange={e => this.onChangeInput(e, 'email')}
            placeholder="your email"
            name="email"
            disabled={!!this.props.initialValues}
            type="email"
            value={email}
          />
          {this.props.initialValues &&
            'To update the email address, please create another alert.'}
        </div>
        <input className="Button" type="submit" value="GO" />
      </form>
    );
  }
}

export default Form;
