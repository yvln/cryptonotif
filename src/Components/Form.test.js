import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';

import { initializeFirebase } from '../firebase';

import Form from './Form';

configure({ adapter: new Adapter() });

describe('src/Components/Form', () => {
  it('should render', () => {
    initializeFirebase();
    const component = renderer.create(
      <Form
        initialValues={{
          data: {
            asset: 'BTC',
            amount: '600',
            action: 'above',
            currency: 'USD',
            email: 'itsme@contact.com',
          },
        }}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render the Edition From when there are initial values', () => {
    const form = mount(
      <Form
        initialValues={{
          data: {
            asset: 'BTC',
            amount: '600',
            action: 'above',
            currency: 'USD',
            email: 'itsme@contact.com',
          },
        }}
      />
    );
    expect(form.find('.FormEdit')).toHaveLength(1);
    expect(form.find('.FormCreate')).toHaveLength(0);
  });

  it('should render the Creation From when there are no initial values', () => {
    const form = mount(<Form />);
    expect(form.find('.FormCreate')).toHaveLength(1);
    expect(form.find('.FormEdit')).toHaveLength(0);
  });
});
