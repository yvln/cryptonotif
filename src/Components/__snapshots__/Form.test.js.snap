// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/Components/Form should render 1`] = `
<form
  className="Form FormEdit"
  onSubmit={[Function]}
>
  <div>
    Alert me when
  </div>
  <div
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    className="react-autosuggest__container"
    role="combobox"
  >
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-controls="react-autowhatever-1"
      autoComplete="off"
      className="react-autosuggest__input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="BTC, ETH, ..."
      type="text"
      value="BTC"
    />
    <div
      className="react-autosuggest__suggestions-container"
      id="react-autowhatever-1"
      role="listbox"
    />
  </div>
  <select
    className="menuAction"
    name="action"
    onChange={[Function]}
    value="above"
  >
    <option
      value="under"
    >
      falls under
    </option>
    <option
      value="above"
    >
      is above
    </option>
  </select>
  <div>
    <input
      name="amount"
      onChange={[Function]}
      type="number"
      value="600"
    />
    <select
      className="menuCurrency"
      name="currency"
      onChange={[Function]}
      value="USD"
    >
      <option
        value="USD"
      >
        USD
      </option>
      <option
        value="EUR"
      >
        EUR
      </option>
    </select>
  </div>
  <div>
    <span>
      to 
    </span>
    <input
      disabled={
        Object {
          "data": Object {
            "action": "above",
            "amount": "600",
            "asset": "BTC",
            "currency": "USD",
            "email": "itsme@contact.com",
          },
        }
      }
      name="email"
      onChange={[Function]}
      placeholder="your email"
      type="email"
      value="itsme@contact.com"
    />
    To update the email address, please create another alert.
  </div>
  <input
    className="Button"
    type="submit"
    value="GO"
  />
</form>
`;
